test(implemented full tests for Payroll by Index module)

lomamech opened the pull request:
- Verified the correct transition of BHIMA system to Payroll by Index mode.
- Added and executed unit tests (client/server) for all core payroll modules:
- functions, grades, holidays, offdays, iprTax, payrollConfig, rubrics, rubricConfig, accountConfig, weekendConfig, employeeConfig.
- Implemented and validated E2E tests for:
- Payroll creation and payment via Multiple Payroll (Indice).
- Proper setup of the Indice Payroll Rubrics and configuration consistency.
- Added integration tests for staffingIndices and staffingIndicesReport modules.
- Verified CRUD operations for payroll-related entities (Grade, Function, Rubric, Configurations).
- Confirmed correct calculation of: Base Index, Worked Days, Daily Index, Adjusted Index, Gross Salary, INSS, IPR, and Net Salary.
- Ensured that all test suites pass locally, including complex payroll scenarios with multiple employees.
- Verified system transition to “Payroll by Index” mode
- Added unit tests for functions
- Added unit and E2E tests for job titles
- Added unit tests for grades
- Added unit tests for holidays
- Added unit tests for offdays
- Added unit tests for IPR tax
- Added unit tests for payroll configuration
- Added unit tests for rubrics
- Added unit tests for rubric configuration
- Added unit tests for account configuration
- Added unit tests for weekend configuration
- Added unit tests for employee configuration
- Added unit tests for multiple payroll processing
- Added unit tests for multiple payroll by index
- Added unit tests for staffing indices
- Updated payroll tests to retrieve the latest enterprise settings directly from the database

lomamech committed:
test(implemented full tests for Payroll by Index module)
- Verified the correct transition of BHIMA system to Payroll by Index mode.
- Added and executed unit tests (client/server) for all core payroll modules:
- functions, grades, holidays, offdays, iprTax, payrollConfig, rubrics, rubricConfig, accountConfig, weekendConfig, employeeConfig.
- Implemented and validated E2E tests for:
- Payroll creation and payment via Multiple Payroll (Indice).
- Proper setup of the Indice Payroll Rubrics and configuration consistency.
- Added integration tests for staffingIndices and staffingIndicesReport modules.
- Verified CRUD operations for payroll-related entities (Grade, Function, Rubric, Configurations).
- Confirmed correct calculation of: Base Index, Worked Days, Daily Index, Adjusted Index, Gross Salary, INSS, IPR, and Net Salary.
- Ensured that all test suites pass locally, including complex payroll scenarios with multiple employees.
- Verified system transition to “Payroll by Index” mode
- Added unit tests for functions
- Added unit and E2E tests for job titles
- Added unit tests for grades
- Added unit tests for holidays
- Added unit tests for offdays
- Added unit tests for IPR tax
- Added unit tests for payroll configuration
- Added unit tests for rubrics
- Added unit tests for rubric configuration
- Added unit tests for account configuration
- Added unit tests for weekend configuration
- Added unit tests for employee configuration
- Added unit tests for multiple payroll processing
- Added unit tests for multiple payroll by index
- Added unit tests for staffing indices
- Updated payroll tests to retrieve the latest enterprise settings directly from the database

jniles commented on test/server-unit/payroll_test_unit/titles.spec.js:
```diff
@@ -0,0 +1,184 @@
+  });
+
+  // ---------------------------------------------------------
+  it('list() should return status 200 with the list of job titles', async () => {
```
It looks like this test is testing expressjs rather than BHIMA.  I'm not sure these kinds of tests are valuable.

Correct me if I am wrong, but this (1) fakes the DB connection, (3) calls controller.list() and then checks if the fake data was returned as JSON.  This is basically what our integration tests do, and I'm not sure what this adds over the integration tests.

jniles commented on test/integration/test_payroll_process_scenarios_2/payroll_index_payment.integration.test.js:
```diff
@@ -0,0 +1,105 @@
+        // Test only if the array contains elements
+        if (employee.otherProfits.length > 0) {
+          employee.otherProfits.forEach(profit => {
+            expect(profit).to.have.property('value');
```
You can simplify both these assertions to be :

```js
expect(profit).to.have.all.keys('label', 'value');
```

I believe this will throw an error if there are more keys than just those two.   If we have more keys that we return, let's check for those here.

jniles reviewed (commented):
@lomamech this is a lot of code and represents a lot of good work.

I've made some comments to try and tighten up the tests.  I think the biggest issues to work on are:

1. I don't really like the use of `setImmediate()` in these tests.  Let's be using `async/await` where possible, even if that means changing the server API slightly.  It will mean all your tests will start to look the same, and that is a good thing.
2. Try to avoid calls that check if something is true (e.g. `expect(x.calledOnce()).to.equal(true)`).  Instead, use `expect(x).to.be.calledOnce()`.  These will make the failures easier to identify in the future.
3. I think your tests for payroll should be stricter - you should state how many records you expect to find, and what properties they should have.  That way, we will know when our internal business logic is violated.

In all, this good work.  Thanks for all your contributions!

jniles committed:
fix(tests): ensure E2E tests pass

This commit fixes some non-specific selectors that were causing crashes
in the end to end tests.

jniles committed:
chore: rm console.log() call

jniles reviewed (approved):
@lomamech this looks good to me.  I've made a few edits here to fix some E2E test errors.  I will merge this PR.

Please confirm that all E2E tests pass on your machine.
